android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java

```diff
@@ -128,18 +140,105 @@ public List<TwoLineItemText> accountParamDataToList() {
         return twoLineItemDataSources;
     }
 
+    public List<TwoLineItemText> programIdToList() {
+        return Arrays.asList(new TwoLineItemText(null, mSolanaInstruction.programId));
+    }
+
     public List<TwoLineItemText> dataToList() {
         return Arrays.asList(new TwoLineItemText(null, Arrays.toString(mSolanaInstruction.data)));
     }
 
+    // Get lamport from decoded data params
+    public String getLamportAmount() {
+        if (isDecodedDataPresent && mSolanaInstruction.decodedData.params != null) {
+            for (SolanaInstructionParam instructionParam : mSolanaInstruction.decodedData.params) {
+                if (instructionParam.name.equalsIgnoreCase(WalletConstants.SOL_LAMPORTS)) {
+                    return instructionParam.value;
+                }
+            }
+        }
+        return "0";
+    }
+
+    public Integer getInstructionType() {
+        if (mInstructionType != null) return mInstructionType;
+        if (isDecodedDataPresent) {
+            mInstructionType = mSolanaInstruction.decodedData.instructionType;
+            return mInstructionType;
+        }
+        return null;
+    }
+
+    public String fromPubKey() {
+        if (mFromPubKey != null) return mFromPubKey;
+        if (isAccountMetaPresent()) {
+            SolanaInstructionAccountParam[] accountParams =
+                    mSolanaInstruction.decodedData.accountParams;
+            for (int i = 0; i < accountParams.length; i++) {
+                if (accountParams[i].name.equalsIgnoreCase(WalletConstants.SOL_DAPP_FROM_ACCOUNT)) {
+                    if (mSolanaInstruction.accountMetas.length > i) {
+                        mFromPubKey = mSolanaInstruction.accountMetas[i].pubkey;
+                    }
+                    return mFromPubKey;
+                }
+            }
+        }
+        return null;
+    }
+
+    public String toPubKey() {
+        if (mToPubKey != null) return mToPubKey;
+        if (isAccountMetaPresent()) {
+            SolanaInstructionAccountParam[] accountParams =
+                    mSolanaInstruction.decodedData.accountParams;
+            for (int i = 0; i < accountParams.length; i++) {
+                if (accountParams[i].name.equalsIgnoreCase(WalletConstants.SOL_DAPP_TO_ACCOUNT)) {
```

Comment:
the part of  `fromPubKey()`  which extracts pubkey is the same as one at `toPubKey()`
the only difference is `WalletConstants.SOL_DAPP_FROM_ACCOUNT` vs `WalletConstants.SOL_DAPP_TO_ACCOUNT`

Is it possible to make something like `String getMetaPubKeyByParam(String paramName)` and use it on both `fromPubKey()` and `toPubKey()` ?

Reply:
oh, `getPubKeyPerParamKey` below does something similar, so can it also   re-use the most general method?